src/main.ts

```diff
@@ -0,0 +1,32 @@
+import { Actor } from 'apify';
+import { log, LogLevel, PlaywrightCrawler } from 'crawlee';
+import { router } from './routes.js';
+import { InputSchema } from './types.js';
+import { validateInput } from './tools.js';
+import configuration from './configuration.js';
+import dataLoader from './data_loader.js';
+
+await Actor.init();
+
+const input = await Actor.getInput<InputSchema>();
+
+validateInput(input);
+if ((input as InputSchema).debug) {
+    log.setLevel(LogLevel.DEBUG);
+}
+
+await configuration.init(input as InputSchema, Actor.getEnv());
+await dataLoader.init(input as InputSchema);
+
+const crawler = new PlaywrightCrawler({
+    requestHandler: router,
+});
+
+// Do not load initial batch after migration
+const startUrls = !dataLoader.initialBatchLoaded
+    ? await dataLoader.getNextBatch()
+    : [];
+
+await crawler.run(startUrls);
```

Comment:
I recommend a simpler batching system that will allow you to remove some of the code connected with state persistence. 

You can have 2 crawlers.
- 1st crawler will be used to collect the links to the KV store records. You can paginate the KV stored and use the`exclusiveStartKey` (https://docs.apify.com/apify-client-js#keyvaluestoreclient) as a uniqueKey in the request queue. You can push these links either to temporary named dataset or to persisted object.
- 2nd crawler loads the HTMLs and renders screenshots. You already have this but you don't need the batches, you can just use RequestList

Reply:
I am not sure that I exactly understand how it should work. Does this solution also solve loading data from a dataset? I just might be missing something.

Reply:
From dataset you can just do that in a single request since it will give you all items at once. KV store is the more tricky one since you have to iterate over 1000.

Reply:
Oh, I also think there was a misunderstanding about the loading from dataset feature. It was mainly meant for datasets that have links to stored HTML as the field, not the HTML itself. So in this way it will be similar to the key value store way.